### PR TITLE
Pre-P0 changes

### PR DESCRIPTION
**Pull request by vlad17**
* auto-infer params when user uses gflags
* infer local dir from git repo name
* added context-based global trial interface
* provide `artifact_directory` interface instead of the file-based save artifact interface
* added automatic logging utilities
* misc bugfixes to make local dir stuff work
* console-based browser for trials added

Tested manually on local directory mode.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -5,22 +5,46 @@
...
+matplotlib.use('Agg')
+import matplotlib.pyplot as plt
+
+remote_dir = "s3://my-track-bucket/myproject"
+proj = track.Project("~/track/myproject", remote_dir)
+most_recent = proj.ids["start_time"].idxmax()
+most_recent_id = proj.ids["trial_id"].iloc[[most_recent]]
+res = proj.results(most_recent_id)
+plt.plot(res["loss"].dropna())
+plt.savefig("loss.png")
```
Comment (richardliaw) [inline README.md:49]: don't actually need matplotlib here right?

File: README.md [hunk 2]
```diff
@@ -5,22 +5,46 @@
 `pip install track` (Not yet). Until we get pypi set up (there's another `track` package...), use
 
 ```
-pip install --upgrade git+git://github.com/richardliaw/track.git#egg=track
+pip install --upgrade git+https://github.com/richardliaw/track.git@master#egg=track
 ```
 
 ## Usage
 
+Report various metrics of interest (if you launch within a git repository `myproject`, the local directory `~/track/myproject` is automatically inferred below).
```
Comment (richardliaw) [inline README.md]: Nit: Can you remove these parentheticals? They're sort of distracting (and can go into a separate example below)

File: README.md [hunk 3]
```diff
@@ -5,22 +5,46 @@
...
+    with track.trial("~/track/myproject", "s3://my-track-bucket/myproject"):
+        model = create_model()
+        for epoch in range(100):
+            model.train()
+            loss = model.get_loss()
+            track.metric(iteration=epoch, loss=loss)
 ```
         
-    
+Inspect existing experiments (from anywhere, if you supply `--remote_dir`)
```
Comment (richardliaw) [inline README.md]: (above comment about parens)

File: track/__init__.py [hunk 4]
```diff
@@ -11,13 +13,24 @@
...
+def trial(log_dir=None,
           upload_dir=None,
           sync_period=None,
           trial_prefix="",
-          param_map=None):
-    """Generates a trial within a with context"""
+          param_map=None,
+          init_logging=True):
+    """
+    Generates a trial within a with context.
```
Comment (richardliaw) [inline track/__init__.py:32]: Can you doc what `init_logging` does?

File: track/__init__.py [hunk 5]
```diff
@@ -33,17 +46,17 @@ def trial(log_dir="~/ray_results/project_name",
...
 
 def metric(*, iteration=None, **kwargs):
     """Applies Trial.metric to the trial in the current context."""
-    _trial.metric(iteration=iteration, **kwargs)
+    return _trial.metric(iteration=iteration, **kwargs)
 
-def artifact(artifact_name, src):
-    """Applies Trial.artifact to the trial in the current context."""
-    _trial.artifact(artifact_name, src)
+def artifact_directory():
```
Comment (richardliaw) [inline track/__init__.py]: `get_dir`?

File: track/autodetect.py [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+    then that is used.
+
+    Otherwise, we just say it's "unknown"
+    """
+    project_name = git_repo()
+    if not project_name and sys.argv:
+        project_name = sys.argv[0]
+    if not project_name:
+        project_name = "unknown"
+    dirpath = os.path.join("~", "track", project_name)
```
Comment (richardliaw) [inline track/autodetect.py]: can you make these constants?

File: track/trial.py [hunk 7]
```diff
@@ -74,16 +127,33 @@ def metric(self, *, iteration=None, **kwargs):
...
+        # unsure if editting the param_map file like this is very kosher
+        self.param_map["trial_completed"] = True
+        # using a constructor for its side effect here (updating the param map)
+        UnifiedLogger(
+            self.param_map, self.data_dir, self.trial_id + "_").close()
 
     def get_result_filename(self):
         return os.path.join(self.data_dir, self.trial_id + "_" + RESULT_SUFFIX)
+
+def _git_hash():
```
Comment (richardliaw) [inline track/trial.py]: can you move this into utilities file like `autodetect`?

File: track/log.py [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+    pass
+
+class StdoutHandler(logging.StreamHandler):
+    """As described by the name"""
+    def __init__(self):
+        super().__init__(sys.stdout)
+
+def init(track_log_handler):
+    """
+    (Re)initialize track's file handler for track package logger.
```
Comment (richardliaw) [inline track/log.py:32]: what exactly is this doing? 

Is the idea that users will log things via 
`logger = logging.getLogger("track")`?

File: README.md (hunk 1 above)
Comment (vlad17) [reply README.md:49]: You do for the thing to save if you're running on a server

File: README.md (hunk 2 above)
Comment (vlad17) [reply README.md]: done

File: README.md (hunk 3 above)
Comment (vlad17) [reply README.md]: done

File: track/__init__.py (hunk 5 above)
Comment (vlad17) [reply track/__init__.py]: how about trial_dir()?

File: track/autodetect.py (hunk 6 above)
Comment (vlad17) [reply track/autodetect.py]: tbh i think constants.py is a bit premature but done

File: track/trial.py (hunk 7 above)
Comment (vlad17) [reply track/trial.py]: done

File: track/log.py (hunk 8 above)
Comment (vlad17) [reply track/log.py:32]: No, they should log using track.debug() (updated readme)

File: track/__init__.py (hunk 4 above)
Comment (vlad17) [reply track/__init__.py:32]: I do in Trial(). we can make good docs later

File: track/trial.py [hunk 9]
```diff
@@ -74,16 +127,18 @@ def metric(self, *, iteration=None, **kwargs):
...
-        shutil.copy(srcpath, destpath)
+    def trial_dir(self):
+        """returns the local file path to the trial's artifact directory"""
+        return self.artifact_dir
 
     def close(self):
         for hook in self._hooks:
             hook.close()
+        # unsure if editting the param_map file like this is very kosher
+        self.param_map["trial_completed"] = True
```
Comment (richardliaw) [inline track/trial.py:138]: super sketch ..

Comment (vlad17) [reply track/trial.py:138]: lol it's OK if you ack the hack

File: track/trials.py [hunk 10]
```diff
@@ -0,0 +1,87 @@
...
+from . import Project
+
+flags.DEFINE_string("local_dir", None,
+                          "the local directory where trials are stored "
+                          "for default behavior see "
+                          "track.autodetect.dfl_local_dir")
+flags.DEFINE_string("remote_dir", None,
+                          "the remote directory where trials are stored")
+
+def compare(c, l, r):
```
Comment (richardliaw) [inline track/trials.py:49]: https://stackoverflow.com/questions/44554614/eval-string-cmp-operator

💯

**Review by richardliaw — APPROVED** [review]
